### PR TITLE
Ensure no NaN in gradient with vector formulation and uniform permittivity

### PR DESCRIPTION
**Pull request by mfschubert**
Current vector field formulations can lead to NaN gradient when the permittivity is uniform. Here, we adjust how the special case of uniform permittivity is handled, which seems to resolve the issue.

A new test is added to `test_grad.py` file, which would have failed previously but now passes.

Tests for several examples are adjusted to run at higher resolution (i.e. higher N) so that they (hopefully) become less sensitive to small changes in the code. Results for reasonable N are essentially unchanged, as seen in uLED efficiency plot below. These can be compared to the figure in #70.

image

### CODE REVIEW

File: src/fmmax/vector.py [hunk 1]
```diff
@@ -189,18 +189,26 @@ def _compute_tangent_field_no_batch(
...
+    # through the vector field calculation.
     gx_is_zero = jnp.all(
         jnp.isclose(grad[..., 0, jnp.newaxis], 0.0), axis=(-3, -2, -1), keepdims=True
     )
     gy_is_zero = jnp.all(
         jnp.isclose(grad[..., 1, jnp.newaxis], 0.0), axis=(-3, -2, -1), keepdims=True
     )
-    grad = jnp.where(gx_is_zero & gy_is_zero, jnp.ones_like(grad), grad)
+    dummy_grad = jnp.broadcast_to(jnp.asarray([1, 0], dtype=complex), grad.shape)
+    grad = jnp.where(gx_is_zero & gy_is_zero, dummy_grad, grad)
```
Comment (smartalecH) [inline src/fmmax/vector.py:211]: The "double where" trick my be more canonical/robust here:

https://jax.readthedocs.io/en/latest/faq.html#gradients-contain-nan-where-using-where

Comment (mfschubert) [reply src/fmmax/vector.py:211]: That's exactly what this is. :-)

We give a dummy gradient to the vector field calculation, so as to avoid it producing NaNs. This is the first `where`. Then, below we manually overwrite the result of that calculation with a safe value (with the second `where`).

Comment (mfschubert) [reply src/fmmax/vector.py:211]: The issue previously was that grad was coming up as nonzero, and so the "safe" value was not used. This happened because the gradient was calculated for a "smoothed" field. Now we reverse the sequence --- first calculate the gradient for the raw permittivity, and then smooth this.

Comment (smartalecH) [reply src/fmmax/vector.py:211]: Got it, thanks!

**Review by smartalecH — APPROVED** [review]
